x/staking/src/proto.rs

```diff
@@ -54,25 +87,88 @@ pub struct Validator {
 }
 
 impl Validator {
-    pub fn abci_validator_update(&self) -> ValidatorUpdate {
-        todo!()
-        // ValidatorUpdate {
-        //     pub_key: Some(self.consensus_pubkey.into()),
-        //     power: self.tendermint_power(),
-        // }
+    pub fn abci_validator_update(&self, power: i64) -> ValidatorUpdate {
+        ValidatorUpdate {
+            pub_key: Some(self.consensus_pubkey.clone().into()),
+            power: self.consensus_power(power),
+        }
+    }
+    pub fn abci_validator_update_zero(&self) -> ValidatorUpdate {
+        self.abci_validator_update(0)
+    }
+
+    pub fn tm_cons_public_key(&self) -> AccAddress {
+        self.consensus_pubkey.get_address()
+    }
+
+    pub fn get_cons_addr(&self) -> AccAddress {
+        // TODO: the other logic that
+        self.consensus_pubkey.get_address()
     }
+
+    pub fn update_status(&mut self, status: BondStatus) {
+        self.status = status;
+    }
+
     pub fn tendermint_power(&self) -> i64 {
         if self.status == BondStatus::Bonded {
             return self.potential_tendermint_power();
         }
         0
     }
+
     pub fn potential_tendermint_power(&self) -> i64 {
-        let amount = self
-            .tokens
-            .amount
+        let amount = self.tokens.amount;
+        let amount = amount / Uint256::from(10u64).pow(6);
+        amount
+            .to_string()
+            .parse::<i64>()
+            .expect("Unexpected conversion error")
+    }
+
+    pub fn consensus_power(&self, power: i64) -> i64 {
+        match self.status {
+            BondStatus::Bonded => self.potential_consensus_power(power),
+            _ => 0,
+        }
+    }
+
+    pub fn potential_consensus_power(&self, power: i64) -> i64 {
+        self.tokens_to_consensus_power(power)
+    }
+
+    pub fn tokens_to_consensus_power(&self, power: i64) -> i64 {
+        let amount = self.tokens.amount;
+        let amount = amount / Uint256::from(power as u64);
+        amount
+            .to_string()
             .parse::<i64>()
-            .expect("Unexpected conversion error");
-        amount / 10i64.pow(6)
+            .expect("Unexpected conversion error")
+    }
+
+    /// GetValidatorsByPowerIndexKey creates the validator by power index.
+    /// Power index is the key used in the power-store, and represents the relative
+    /// power ranking of the validator.
+    /// VALUE: validator operator address ([]byte)
+    pub fn key_by_power_index_key(&self, power_reduction: i64) -> Vec<u8> {
```

Comment:
Nit: could we use an array instead of a vec here i.e. does this return a fixed length?

Reply:
@joneskm We can use an array if ValAddress has constant length. As I understand the length of the address is not static but restricted by the constant https://github.com/rumos-io/gears/blob/main/tendermint/src/types/proto/crypto.rs#L2. 

Am I correct?

I updated the realization of method `get_unbonding_delegation_time_key` https://github.com/rumos-io/gears/pull/81/commits/9fa274485ac5de16040803ddcf7f8d4fe6b06b80#diff-d215ed340a59cbde946a9a09538396a743896c746fe271374eb0e8a71c2e3635R404

Reply:
Ah OK. The ValAddress is not a constant length so ignore my comment.